src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs

```diff
@@ -56,222 +57,162 @@ public CbcAuthenticatedEncryptor(Secret keyDerivationKey, BCryptAlgorithmHandle
+                    UnsafeBufferUtil.BlockCopy(from: pbIV, to: pbClonedIV2, byteCount: _symmetricAlgorithmBlockSizeInBytes);
+
+                    // Perform the actual decryption
+                    fixed (byte* pbBuffer = buffer)
+                    {
+                        byte dummy;
+                        ntstatus = UnsafeNativeMethods.BCryptDecrypt(
+                            hKey: decryptionSubkeyHandle,
+                            pbInput: pbEncryptedData,
+                            cbInput: (uint)cbEncryptedDataLength,
+                            pPaddingInfo: null,
+                            pbIV: pbClonedIV2,
+                            cbIV: _symmetricAlgorithmBlockSizeInBytes,
+                            pbOutput: (buffer.Length > 0) ? pbBuffer : &dummy,
+                            cbOutput: (uint)buffer.Length,
+                            pcbResult: out var dwActualDecryptedByteCount,
+                            dwFlags: BCryptEncryptFlags.BCRYPT_BLOCK_PADDING);
+
+                        UnsafeNativeMethods.ThrowExceptionForBCryptStatus(ntstatus);
+
+                        // Advance the writer by the number of bytes actually written
+                        destination.Advance(checked((int)dwActualDecryptedByteCount));
+                    }
                 }
-
-                ptr += _hmacAlgorithmDigestLengthInBytes;
-                CryptoUtil.Assert(ptr - pbRetVal == retVal.Length, "ptr - pbRetVal == retVal.Length");
+            }
+            finally
+            {
+                // Buffer contains sensitive key material; delete.
+                UnsafeBufferUtil.SecureZeroMemory(pbTempSubkeys, cbTempSubkeys);
             }
         }
-
-        // retVal := { version || chainingMode || symAlgKeySize || symAlgBlockSize || hmacAlgKeySize || hmacAlgDigestSize || E("") || MAC("") }.
-        return retVal;
     }
 
-    protected override byte[] DecryptImpl(byte* pbCiphertext, uint cbCiphertext, byte* pbAdditionalAuthenticatedData, uint cbAdditionalAuthenticatedData)
+    public byte[] Decrypt(ArraySegment<byte> ciphertext, ArraySegment<byte> additionalAuthenticatedData)
     {
-        // Argument checking - input must at the absolute minimum contain a key modifier, IV, and MAC
-        if (cbCiphertext < checked(KEY_MODIFIER_SIZE_IN_BYTES + _symmetricAlgorithmBlockSizeInBytes + _hmacAlgorithmDigestLengthInBytes))
+        ciphertext.Validate();
+        additionalAuthenticatedData.Validate();
+
+        var outputSize = ciphertext.Count - (int)(KEY_MODIFIER_SIZE_IN_BYTES + _symmetricAlgorithmBlockSizeInBytes + _hmacAlgorithmDigestLengthInBytes);
+        if (outputSize <= 0)
         {
             throw Error.CryptCommon_PayloadInvalid();
         }
 
-        // Assumption: pbCipherText := { keyModifier | IV | encryptedData | MAC(IV | encryptedPayload) }
+#if NET
+        byte[] rentedBuffer = null!;
```

Comment:
As a general rule, we shouldn't use the null-forgiving operator when initializing variables/fields/properties, unless we're intentionally lying about the nullability of a public property that's always non-null by the time the user sees it, and even that's super questionable. It's better to save the null-forgiving until later in the code when we know for sure the variable/field/property must not be null. In this case, the only other place we use it we have a `if (rentedBuffer is not null)` anyways.

```suggestion
        byte[]? rentedBuffer = null;
```

This is really making me wish C# had macros though. I wonder if we should just have another overload to RefPooledArrayBufferWriter that takes an initial size other than a stackalloced Span. Not only would it simplify a lot of this repeated code a little bit, this would ensure the `RefPooledArrayBufferWriter` is always be responsible for renting and returning arrays to the pool.